localization/ndt_scan_matcher/src/ndt_scan_matcher_core.cpp

```diff
@@ -841,47 +839,54 @@ std::array<double, 36> NDTScanMatcher::estimate_covariance(
   multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(ndt_result.pose));
   multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(initial_pose_matrix));
 
-  // multiple searches
-  for (const auto & pose_offset :
-       param_.covariance.covariance_estimation.initial_pose_offset_model) {
-    const Eigen::Vector2d rotated_pose_offset_2d = rot * pose_offset;
-
-    Eigen::Matrix4f sub_initial_pose_matrix(Eigen::Matrix4f::Identity());
-    sub_initial_pose_matrix = ndt_result.pose;
-    sub_initial_pose_matrix(0, 3) += static_cast<float>(rotated_pose_offset_2d.x());
-    sub_initial_pose_matrix(1, 3) += static_cast<float>(rotated_pose_offset_2d.y());
-
-    auto sub_output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
-    ndt_ptr_->align(*sub_output_cloud, sub_initial_pose_matrix);
-    const Eigen::Matrix4f sub_ndt_result = ndt_ptr_->getResult().pose;
-
-    const Eigen::Vector2d sub_ndt_pose_2d = sub_ndt_result.topRightCorner<2, 1>().cast<double>();
-    mean += sub_ndt_pose_2d;
-    ndt_pose_2d_vec.emplace_back(sub_ndt_pose_2d);
-
-    multi_ndt_result_msg.poses.push_back(matrix4f_to_pose(sub_ndt_result));
-    multi_initial_pose_msg.poses.push_back(matrix4f_to_pose(sub_initial_pose_matrix));
+  std::vector<double> initial_pose_offset_model_x;
+  std::vector<double> initial_pose_offset_model_y;
+  const size_t offset_size =
+    param_.covariance.covariance_estimation.initial_pose_offset_model.size();
+  for (size_t i = 0; i < offset_size; i++) {
+    initial_pose_offset_model_x.push_back(
+      param_.covariance.covariance_estimation.initial_pose_offset_model[i].x());
+    initial_pose_offset_model_y.push_back(
+      param_.covariance.covariance_estimation.initial_pose_offset_model[i].y());
   }
 
-  // calculate the covariance matrix
-  mean /= n;
-  Eigen::Matrix2d pca_covariance = Eigen::Matrix2d::Zero();
-  for (const auto & temp_ndt_pose_2d : ndt_pose_2d_vec) {
-    const Eigen::Vector2d diff_2d = temp_ndt_pose_2d - mean;
-    pca_covariance += diff_2d * diff_2d.transpose();
+  if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::LAPLACE_APPROXIMATION) {
+    return pclomp::estimate_xy_covariance_by_Laplace_approximation(ndt_result.hessian);
+  } else if (
+    param_.covariance.covariance_estimation.covariance_estimation_type ==
+    CovarianceEstimationType::MULTI_NDT) {
+    const std::vector<Eigen::Matrix4f> poses_to_search = pclomp::propose_poses_to_search(
+      ndt_result, initial_pose_offset_model_x, initial_pose_offset_model_y);
+    const pclomp::ResultOfMultiNdtCovarianceEstimation result_of_multi_ndt_covariance_estimation =
+      estimate_xy_covariance_by_multi_ndt(ndt_result, ndt_ptr_, poses_to_search);
+    for (const auto & sub_initial_pose_matrix : poses_to_search) {
+      auto sub_output_cloud = std::make_shared<pcl::PointCloud<PointSource>>();
+      ndt_ptr_->align(*sub_output_cloud, sub_initial_pose_matrix);
```

Comment:
Is this "align" necessary?

The struct `pclomp::ResultOfMultiNdtCovarianceEstimation` already has `std::vector<NdtResult> ndt_results`.

https://github.com/tier4/ndt_omp/blob/ef654bdb1a3fe0bae6f3247b4959e230e63ba429/include/estimate_covariance/estimate_covariance.hpp#L9-L14

Reply:
Sure, maybe. I'll fix it.